modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java

```diff
@@ -86,23 +86,24 @@ protected void doDispatch(
 			WebKeys.THEME_DISPLAY);
 
 		try {
-			_generate(
+			ObjectDefinition fdsEntryObjectDefinition = _generate(
 				themeDisplay.getCompanyId(), themeDisplay.getLocale(),
 				themeDisplay.getUserId());
+
+			renderRequest.setAttribute(
+				FDSViewsWebKeys.FDS_VIEWS_DISPLAY_CONTEXT,
+				new FDSViewsDisplayContext(
+					_cetManager, fdsEntryObjectDefinition, renderRequest,
+					renderResponse, _serviceTrackerList));
 		}
 		catch (Exception exception) {
 			_log.error(exception);
 		}
 
-		renderRequest.setAttribute(
-			FDSViewsWebKeys.FDS_VIEWS_DISPLAY_CONTEXT,
-			new FDSViewsDisplayContext(
-				_cetManager, renderRequest, _serviceTrackerList));
-
 		super.doDispatch(renderRequest, renderResponse);
 	}
 
-	private synchronized void _generate(
+	private synchronized ObjectDefinition _generate(
```

Comment:
Naming and scoping of concerns doesn't seem right here. A specific instance of `ObjectDefinition` is weird as returned value from a method that generates all persistance. What will happen when we will want [permissions action on Data Set Views](https://liferay.atlassian.net/browse/LPS-183894), and we need `fdsViewObjectDefinition` in context?

We may want to do something like:
```js
ObjectDefinition fdsEntryObjectDefinition = _getFDSEntryObjectDefinition();

_generate(fdsEntryObjectDefinition, ...);

// forward fdsEntryObjectDefinition to context
```

... but I'm not sure if Objects API would support this.

Alternatively, we could:
- squash `_generate` method inline in `doDispatch`, at the cost of one indent  - or -
- set context attribute within `_generate`

Reply:
Passing in `fdsEntryObjectDefinition` in `_generate(fdsEntryObjectDefinition, ...);` won't work since `_getFDSEntryObjectDefinition()` might be null for 2 different threads and both attempt to create new objects.

`_generate` also has to be a separate method so it can be `synchronized`.

I'll think more about how to restructure this.